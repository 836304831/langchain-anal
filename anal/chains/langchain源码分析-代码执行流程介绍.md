# chains源码分析-关键点说明

## chains流程&类介绍
### 流程&功能介绍

以LLMChain类的使用流程为例，LLMChain的流程中除了prompt相关的类，还涉及到类LLMChain,其基类Chain,以及OpenAI类。
从LLChainl的实例出发，依次经过Chain.__call__(), LLMChain._call()及OpenAI.generate_prompt()等方法。完整一次模型的调用。
其流程图如下:
![avatar](./imgs/LLMChain源码流程介绍.png)
LLMChain源码流程介绍图

### 各chains类图
几乎所有的chains类都包含一个Chain基类，该基类包括了属性:
- input_keys: 输入的变量，包括了用于填充prompt模板的输入变量和用户输入的其他变量，但必须包含所有的prompt中的变量。
- output_keys: chains的输出变量，默认是text, 用户可以指定输出名称，便于在各个chains之间交互。
- memory: 它在每个链的开头和结尾都被调用。在开始时，内存加载变量并在链中传递它们。最后，它保存所有返回的变量。
- callbacks: 了解chains调用过程中的状态。回调处理程序在对链的调用的整个生命周期中被调用， 从onchain_start开始，以onchain_end或onchain_error结束。 每个自定义链都可以选择性地调用其他回调方法
- tags: 这些标记与当前链相关联，并作为参数传递给“callbacks”中定义的处理程序。您可以使用这些标记来识别链的特定实例及其用例。
- verbose: 是否打印中间的日志到控制台
- metadata: 此元数据与当前链链的每次调用相关联，并作为参数传递给“callbacks”中定义的处理程序。您可以使用这些来识别链的特定实例及其用例。

同样的，该基类也包含了主要的通用方法：
- _call(): 与acall功能相同，_call()使用__call__()开始调用。acall是异步调用。
- acall(): 同上
- apply(): 支持多输入的调用，相当于多次调用_call()
- arun(): 功能同_call()， 
- run(): 同arun()
- prep_inputs(): 将用户的输入形成{key: value}，方便后续填充promptTemplate内容;验证是否满足prompt模板的数据填充
- prep_outputs(): 验证输出结果; 根据参数决定是否需要将输入和输出一起返回

当前的chains对应，会重载_call等方法，针对自身chain对应的特点进行特殊处理。返回结果统一都是dict。保证流程处理的一致性。

![avatar](./imgs_classes/llm.png)
LLMChain类图

![avatar](./imgs_classes/llm_requests.png)
LLMRequestsChain类图

![avatar](./imgs_classes/mapreduce.png)
MapReduceChain类图

![avatar](./imgs_classes/moderation.png)
OpenAIModerationChain类图

![avatar](./imgs_classes/transform.png)
TransformChain类图

## chains pipeline介绍

### chains pipeline说明

chains pipeline是将多个chains对应串联在一起，单个的chains可以看成是某个功能的实现，而pipeline是通过将各个功能串联起来，使用上一个chains的输出作为下一个chain对应的输入完整复杂完整的任务。包括两种：
- 简单的单输入单输出chain pipeline(只允许单个输入，单个输出，顺序链)，对应的类为：SimpleSequentialChain
- 多输入输出的顺序链&指定输入输出的chain pipeline, 对应的类为：SequentialChain

![avatar](./imgs/sequential流程介绍.png)
sequential流程介绍


### chains pipeline类图
![avatar](./imgs_classes/sequential.png)
SequentialChain&SimpleSequentialChain类图

