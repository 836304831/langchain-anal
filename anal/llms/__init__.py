# -*- encoding: utf-8 -*-
"""
@author: acedar  
@time: 2023/7/17 21:24
@file: __init__.py.py 
"""
