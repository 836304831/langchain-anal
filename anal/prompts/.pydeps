[pydeps]
verbose = 0
pylib = False
max_module_depth = 10

max_bacon = 10
start_color= 180

exclude_exact =
    langchain.utilities
    langchain.utils
    langchain.llms.utils

exclude =
    os
    re
    sys
    collections
    __future__
    requests
    urllib3
    jinja2
    yaml
    langchain.base_language
    langchain.schema
    langchain.sql_database
    langchain.vectorstores
    langchain.cache
    langchain.output_parsers
    langchain.callbacks
    langchain.agents
    langchain.tools
    langchain.llms
    langchain.utilities
    langchain.llms
    langchain.memory
    langchain.utilities
    langchain.chains
    langchain.document_loaders
    langchain.chat_models
    langchain.docstore
    langchain.retrievers
    langchain.embeddings
    langchain.graphs


