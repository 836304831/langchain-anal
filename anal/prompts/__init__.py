# -*- encoding: utf-8 -*-
"""
@author: acedar  
@time: 2023/7/15 15:48
@file: __init__.py.py 
"""
