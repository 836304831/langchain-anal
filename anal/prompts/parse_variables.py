# -*- coding: utf-8 -*-
# ----------------------------
# @Time    : 2023/7/15 11:10
# @Author  : acedar
# @FileName: parse_variables.py
# ----------------------------

from jinja2 import Environment, meta
from string import Formatter


from string import Formatter
from typing import Any, List, Mapping, Sequence, Union


class StrictFormatter(Formatter):
    """A subclass of formatter that checks for extra keys."""

    def check_unused_args(
        self,
        used_args: Sequence[Union[int, str]],
        args: Sequence,
        kwargs: Mapping[str, Any],
    ) -> None:
        """Check to see if extra parameters are passed."""
        extra = set(kwargs).difference(used_args)
        if extra:
            raise KeyError(extra)

    def vformat(
        self, format_string: str, args: Sequence, kwargs: Mapping[str, Any]
    ) -> str:
        """Check that no arguments are provided."""
        if len(args) > 0:
            raise ValueError(
                "No arguments should be provided, "
                "everything should be passed as keyword arguments."
            )
        return super().vformat(format_string, args, kwargs)

    def validate_input_variables(
        self, format_string: str, input_variables: List[str]
    ) -> None:
        dummy_inputs = {input_variable: "foo" for input_variable in input_variables}
        super().format(format_string, **dummy_inputs)

formatter = StrictFormatter()

def variables_from_template(template, **kwargs):
    if "template_format" in kwargs and kwargs["template_format"] == "jinja2":
        # Get the variables for the template
        env = Environment()
        ast = env.parse(template)
        variables = meta.find_undeclared_variables(ast)
    else:
        variables = {
            v for _, v, _, _ in Formatter().parse(template) if v is not None
        }
    return variables


template_string = """Translate the text \
that is delimited by triple backticks \
into a style that is {style}. \
text: ```{text}```
"""

variables = variables_from_template(template_string)

print(variables)

# from langchain.formatting import formatter


kwargs = {
    "style": """American English \
                in a calm and respectful tone
                """,
    "text": """Arrr, I be fuming that me blender lid \
                flew off and splattered me kitchen walls \
                with smoothie! And to make matters worse, \
                the warranty don't cover the cost of \
                cleaning up me kitchen. I need yer help \
                right now, matey!"""
}

result = formatter.format(template_string, **kwargs)
print("result:\n", result)
