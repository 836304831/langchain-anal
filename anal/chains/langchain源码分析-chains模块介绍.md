
# chains说明
chains对组件的一系列调用，因为在很多场景下，一个完成的功能需要拆分成多个组件调用，将多个组件组合在一起形成完整的pipleline。
组件的调用可以理解为是单个功能的实现，可以保证功能的灵活性，在很多的通用场景下都能使用，有利于功能的复用。
chains的调用，以完整任务为单位，贴合实际应用。
chains模块的功能整体分为两部分: 
- 一部分是单个功能实现的chain，是对llm和prompt的封装。
- 另一部分是将每个chains组合在一起，形成一个完整的pipeline。

## chains模块介绍
- LLMChain: 对大模型调用的功能，输出内容和格式由用户输入指定
- TransformChain: 对chains之间的输入和输出进行处理，便于chains之间进行数据传输。支持自定义的转换函数。
- OpenAIModerationChain: 通过审核端点传递输入
- MapReduceChain: 将大文档拆分，拆分的每一部分发送给发送到LLM，然后将结果组合返回。
- LLMRequestsChain: 解析url中的结果，获取对应结果发送给LLM

### LLMChain使用例子
```python
from langchain import PromptTemplate, OpenAI, LLMChain

prompt_template = "列举5个描述场景 {scene} 特点的词"
prompt = PromptTemplate(
    input_variables=["scene"],
    template=prompt_template
)


llm = OpenAI(temperature=0)

# 核心要素： 模型，prompt模板
llm_chain = LLMChain(
    llm=llm,
    prompt=prompt
)

print(llm_chain.input_keys)

# 调用call方法
# 返回输入和输出
output = llm_chain({"scene": "长江"})
print(output)

output = llm_chain.run({"scene": "矿泉水"})
print(output)

# 参数指定，只返回输出
output = llm_chain({"scene": "公园"}, return_only_outputs=True)
print(output)
# 参数指定，只返回输出, 不使用字典方式输入(对后续chain是有影响的)
# output = llm_chain("书包", return_only_outputs=True)
# print(output)

# apply方法: 多输入
input_list = [
    {"scene": "房子"},
    {"scene": "图书馆"}
]
# llm_chain.apply(input_list)
```
### TransformChain使用例子
```python

from langchain.chains import TransformChain, LLMChain, SimpleSequentialChain
from langchain.llms import OpenAI
from langchain.prompts import PromptTemplate
import json

# transformer chain
def select_feature(inputs):
    print("inputs:", inputs)
    feature_str = inputs["multi_feature"]
    single_feature = feature_str.split("*$*")[0]
    return {"single_feature": single_feature}

transfor_chain = TransformChain(
    input_variables=["multi_feature"], 
    output_variables=["single_feature"],
    transform=select_feature,)

output = transfor_chain({'scene': '白云', 'multi_feature': '云朵*$*柔和*$*悠扬'}, return_only_outputs=True)
print(output)
```

## chains pipeline介绍
- SimpleSequentialChain: 每个步骤都有一个单一的输入/输出，并且一个步骤的输出是下一步的输入
- SequentialChain: 多输入输出的顺序链, 难点: 多输入多输出，如何进行变量的映射。上一个链的多输出，如何精准的传递给下一个链

### SimpleSequentialChain 完整功能使用列子
```python
from langchain.llms import OpenAI
from langchain.chains import LLMChain
from langchain.prompts import PromptTemplate
from langchain.chains import SimpleSequentialChain

llm = OpenAI(temperature=0)
template = """ 你是一名小学语文老师，给定一个场景, 使用2-3个词描述其特点：
场景: {scene}
场景特点: """
prompt_template = PromptTemplate(input_variables=["scene"], template=template)
scene_chain = LLMChain(llm=llm, prompt=prompt_template)

llm = OpenAI(temperature=.7)
template = """你是一名小学语文老师，根据给定场景的特点，请造句，要求使用这些场景特点的词,
特点：{feature}
造句结果: """
prompt_template = PromptTemplate(input_variables=["feature"], template=template)
sentence_chain = LLMChain(llm=llm, prompt=prompt_template)

overall_chain = SimpleSequentialChain(chains=[scene_chain, sentence_chain], verbose=True)
review = overall_chain.run("沙滩")
```

### SequentialChain使用例子
```python
llm = OpenAI(temperature=.7)
template = """ 你是一名小学语文老师，给定一个场景, 使用2-3个词描述其特点：
场景: {scene}
场景特点: """
prompt_template = PromptTemplate(input_variables=["scene"], template=template)
scene_chain = LLMChain(llm=llm, prompt=prompt_template, output_key="feature")

llm = OpenAI(temperature=.7)
template = """你是一名小学语文老师，根据给定场景的特点，请造句，要求使用这些场景特点的词,
特点：{feature}
造句结果: """
prompt_template = PromptTemplate(input_variables=["feature"], template=template)
sentence_chain = LLMChain(llm=llm, prompt=prompt_template, output_key="sentence")

llm = OpenAI(temperature=.7)
template = """你是一名小学语文老师，根据给定场景的特点，一次给出这些特点词的反义词,
特点：{feature}
反义词为: """
prompt_template = PromptTemplate(input_variables=["feature"], template=template)
antonym_chain = LLMChain(llm=llm, prompt=prompt_template, output_key="antonym")


from langchain.chains import SequentialChain

overall_chain = SequentialChain(
    chains=[scene_chain, sentence_chain, antonym_chain],
    input_variables=["scene"],
    # Here we return multiple variables
    output_variables=["feature", "sentence", "antonym"],
    verbose=True)
overall_chain({"scene":"白云"})
```

### 带有转换链的pipeline例子
```python
from langchain.chains import TransformChain, LLMChain, SimpleSequentialChain
from langchain.llms import OpenAI
from langchain.prompts import PromptTemplate
import json

# first chain
llm = OpenAI(temperature=.7)
template = """ 你是一名小学语文老师，给定一个场景, 使用2-3个词描述其特点, 词时间使用*$*分割：
场景: {scene}
场景特点: """
prompt_template = PromptTemplate(input_variables=["scene"], template=template)
scene_chain = LLMChain(llm=llm, prompt=prompt_template, 
                          output_key="multi_feature")


# transformer chain
def select_feature(inputs):
    print("inputs:", inputs)
    feature_str = inputs["multi_feature"]
    single_feature = feature_str.split("*$*")[0]
    return {"single_feature": single_feature}

transfor_chain = TransformChain(
    input_variables=["multi_feature"], 
    output_variables=["single_feature"],
    transform=select_feature,)

# third chain
template = """你是一名小学语文老师，根据给定场景的特点造句，
特点：{single_feature}
造句结果: """
prompt_template = PromptTemplate(input_variables=["single_feature"], template=template)
llm = OpenAI(temperature=.7)


sentence_chain = LLMChain(llm=llm, prompt=prompt_template, 
                          output_key="sentence")

from langchain.chains import SequentialChain

overall_chain = SequentialChain(
    chains=[scene_chain, transfor_chain, sentence_chain],
    input_variables=["scene"],
    # Here we return multiple variables
    output_variables=["multi_feature", "single_feature", "sentence"],
    verbose=True)
overall_chain({"scene":"白云"})
```
