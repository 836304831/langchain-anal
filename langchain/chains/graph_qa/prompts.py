# flake8: noqa
from langchain.prompts.prompt import PromptTemplate

_DEFAULT_ENTITY_EXTRACTION_TEMPLATE = """Extract all entities from the following text. As a guideline, a proper noun is generally capitalized. You should definitely extract all names and places.

Return the output as a single comma-separated list, or NONE if there is nothing of note to return.

EXAMPLE
i'm trying to improve Langchain's interfaces, the UX, its integrations with various products the user might want ... a lot of stuff.
Output: Langchain
END OF EXAMPLE

EXAMPLE
i'm trying to improve Langchain's interfaces, the UX, its integrations with various products the user might want ... a lot of stuff. I'm working with Sam.
Output: Langchain, Sam
END OF EXAMPLE

Begin!

{input}
Output:"""
ENTITY_EXTRACTION_PROMPT = PromptTemplate(
    input_variables=["input"], template=_DEFAULT_ENTITY_EXTRACTION_TEMPLATE
)

_DEFAULT_GRAPH_QA_TEMPLATE = """Use the following knowledge triplets to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer.

{context}

Question: {question}
Helpful Answer:"""
GRAPH_QA_PROMPT = PromptTemplate(
    template=_DEFAULT_GRAPH_QA_TEMPLATE, input_variables=["context", "question"]
)

CYPHER_GENERATION_TEMPLATE = """Task:Generate Cypher statement to query a graph database.
Instructions:
Use only the provided relationship types and properties in the schema.
Do not use any other relationship types or properties that are not provided.
Schema:
{schema}
Note: Do not include any explanations or apologies in your responses.
Do not respond to any questions that might ask anything else than for you to construct a Cypher statement.
Do not include any text except the generated Cypher statement.

The question is:
{question}"""
CYPHER_GENERATION_PROMPT = PromptTemplate(
    input_variables=["schema", "question"], template=CYPHER_GENERATION_TEMPLATE
)

NEBULAGRAPH_EXTRA_INSTRUCTIONS = """
Instructions:

First, generate cypher then convert it to NebulaGraph Cypher dialect(rather than standard):
1. it requires explicit label specification only when referring to node properties: v.`Foo`.name
2. note explicit label specification is not needed for edge properties, so it's e.name instead of e.`Bar`.name
3. it uses double equals sign for comparison: `==` rather than `=`
For instance:
```diff
< MATCH (p:person)-[e:directed]->(m:movie) WHERE m.name = 'The Godfather II'
< RETURN p.name, e.year, m.name;
---
> MATCH (p:`person`)-[e:directed]->(m:`movie`) WHERE m.`movie`.`name` == 'The Godfather II'
> RETURN p.`person`.`name`, e.year, m.`movie`.`name`;
```\n"""

NGQL_GENERATION_TEMPLATE = CYPHER_GENERATION_TEMPLATE.replace(
    "Generate Cypher", "Generate NebulaGraph Cypher"
).replace("Instructions:", NEBULAGRAPH_EXTRA_INSTRUCTIONS)

NGQL_GENERATION_PROMPT = PromptTemplate(
    input_variables=["schema", "question"], template=NGQL_GENERATION_TEMPLATE
)

KUZU_EXTRA_INSTRUCTIONS = """
Instructions:

Generate statement with Kùzu Cypher dialect (rather than standard):
1. do not use `WHERE EXISTS` clause to check the existence of a property because Kùzu database has a fixed schema.
2. do not omit relationship pattern. Always use `()-[]->()` instead of `()->()`.
3. do not include any notes or comments even if the statement does not produce the expected result.
```\n"""

KUZU_GENERATION_TEMPLATE = CYPHER_GENERATION_TEMPLATE.replace(
    "Generate Cypher", "Generate Kùzu Cypher"
).replace("Instructions:", KUZU_EXTRA_INSTRUCTIONS)

KUZU_GENERATION_PROMPT = PromptTemplate(
    input_variables=["schema", "question"], template=KUZU_GENERATION_TEMPLATE
)

GREMLIN_GENERATION_TEMPLATE = CYPHER_GENERATION_TEMPLATE.replace("Cypher", "Gremlin")

GREMLIN_GENERATION_PROMPT = PromptTemplate(
    input_variables=["schema", "question"], template=GREMLIN_GENERATION_TEMPLATE
)

CYPHER_QA_TEMPLATE = """You are an assistant that helps to form nice and human understandable answers.
The information part contains the provided information that you must use to construct an answer.
The provided information is authoritative, you must never doubt it or try to use your internal knowledge to correct it.
Make the answer sound as a response to the question. Do not mention that you based the result on the given information.
If the provided information is empty, say that you don't know the answer.
Information:
{context}

Question: {question}
Helpful Answer:"""
CYPHER_QA_PROMPT = PromptTemplate(
    input_variables=["context", "question"], template=CYPHER_QA_TEMPLATE
)

SPARQL_INTENT_TEMPLATE = """Task: Identify the intent of a prompt and return the appropriate SPARQL query type.
You are an assistant that distinguishes different types of prompts and returns the corresponding SPARQL query types.
Consider only the following query types:
* SELECT: this query type corresponds to questions
* UPDATE: this query type corresponds to all requests for deleting, inserting, or changing triples
Note: Be as concise as possible.
Do not include any explanations or apologies in your responses.
Do not respond to any questions that ask for anything else than for you to identify a SPARQL query type.
Do not include any unnecessary whitespaces or any text except the query type, i.e., either return 'SELECT' or 'UPDATE'.

The prompt is:
{prompt}
Helpful Answer:"""
SPARQL_INTENT_PROMPT = PromptTemplate(
    input_variables=["prompt"], template=SPARQL_INTENT_TEMPLATE
)

SPARQL_GENERATION_SELECT_TEMPLATE = """Task: Generate a SPARQL SELECT statement for querying a graph database.
For instance, to find all email addresses of John Doe, the following query in backticks would be suitable:
```
PREFIX foaf: <http://xmlns.com/foaf/0.1/>
SELECT ?email
WHERE {{
    ?person foaf:name "John Doe" .
    ?person foaf:mbox ?email .
}}
```
Instructions:
Use only the node types and properties provided in the schema.
Do not use any node types and properties that are not explicitly provided.
Include all necessary prefixes.
Schema:
{schema}
Note: Be as concise as possible.
Do not include any explanations or apologies in your responses.
Do not respond to any questions that ask for anything else than for you to construct a SPARQL query.
Do not include any text except the SPARQL query generated.

The question is:
{prompt}"""
SPARQL_GENERATION_SELECT_PROMPT = PromptTemplate(
    input_variables=["schema", "prompt"], template=SPARQL_GENERATION_SELECT_TEMPLATE
)

SPARQL_GENERATION_UPDATE_TEMPLATE = """Task: Generate a SPARQL UPDATE statement for updating a graph database.
For instance, to add 'jane.doe@foo.bar' as a new email address for Jane Doe, the following query in backticks would be suitable:
```
PREFIX foaf: <http://xmlns.com/foaf/0.1/>
INSERT {{
    ?person foaf:mbox <mailto:jane.doe@foo.bar> .
}}
WHERE {{
    ?person foaf:name "Jane Doe" .
}}
```
Instructions:
Make the query as short as possible and avoid adding unnecessary triples.
Use only the node types and properties provided in the schema.
Do not use any node types and properties that are not explicitly provided.
Include all necessary prefixes.
Schema:
{schema}
Note: Be as concise as possible.
Do not include any explanations or apologies in your responses.
Do not respond to any questions that ask for anything else than for you to construct a SPARQL query.
Return only the generated SPARQL query, nothing else.

The information to be inserted is:
{prompt}"""
SPARQL_GENERATION_UPDATE_PROMPT = PromptTemplate(
    input_variables=["schema", "prompt"], template=SPARQL_GENERATION_UPDATE_TEMPLATE
)

SPARQL_QA_TEMPLATE = """Task: Generate a natural language response from the results of a SPARQL query.
You are an assistant that creates well-written and human understandable answers.
The information part contains the information provided, which you can use to construct an answer.
The information provided is authoritative, you must never doubt it or try to use your internal knowledge to correct it.
Make your response sound like the information is coming from an AI assistant, but don't add any information.
Information:
{context}

Question: {prompt}
Helpful Answer:"""
SPARQL_QA_PROMPT = PromptTemplate(
    input_variables=["context", "prompt"], template=SPARQL_QA_TEMPLATE
)
