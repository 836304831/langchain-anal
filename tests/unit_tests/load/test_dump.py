"""Test for Serializable base class"""

from typing import Any, Dict

import pytest

from langchain.callbacks.tracers.langchain import LangChainTracer
from langchain.chains.llm import LLMChain
from langchain.chat_models.openai import ChatOpenAI
from langchain.llms.openai import OpenAI
from langchain.load.dump import dumps
from langchain.load.serializable import Serializable
from langchain.prompts.chat import ChatPromptTemplate, HumanMessagePromptTemplate
from langchain.prompts.prompt import PromptTemplate


class Person(Serializable):
    secret: str

    you_can_see_me: str = "hello"

    @property
    def lc_serializable(self) -> bool:
        return True

    @property
    def lc_secrets(self) -> Dict[str, str]:
        return {"secret": "SECRET"}

    @property
    def lc_attributes(self) -> Dict[str, str]:
        return {"you_can_see_me": self.you_can_see_me}


class SpecialPerson(Person):
    another_secret: str

    another_visible: str = "bye"

    # Gets merged with parent class's secrets
    @property
    def lc_secrets(self) -> Dict[str, str]:
        return {"another_secret": "ANOTHER_SECRET"}

    # Gets merged with parent class's attributes
    @property
    def lc_attributes(self) -> Dict[str, str]:
        return {"another_visible": self.another_visible}


class NotSerializable:
    pass


def test_person(snapshot: Any) -> None:
    p = Person(secret="hello")
    assert dumps(p, pretty=True) == snapshot
    sp = SpecialPerson(another_secret="Wooo", secret="Hmm")
    assert dumps(sp, pretty=True) == snapshot


@pytest.mark.requires("openai")
def test_serialize_openai_llm(snapshot: Any) -> None:
    llm = OpenAI(
        model="davinci",
        temperature=0.5,
        openai_api_key="hello",
        # This is excluded from serialization
        callbacks=[LangChainTracer()],
    )
    llm.temperature = 0.7  # this is reflected in serialization
    assert dumps(llm, pretty=True) == snapshot


@pytest.mark.requires("openai")
def test_serialize_llmchain(snapshot: Any) -> None:
    llm = OpenAI(model="davinci", temperature=0.5, openai_api_key="hello")
    prompt = PromptTemplate.from_template("hello {name}!")
    chain = LLMChain(llm=llm, prompt=prompt)
    assert dumps(chain, pretty=True) == snapshot


@pytest.mark.requires("openai")
def test_serialize_llmchain_env() -> None:
    llm = OpenAI(model="davinci", temperature=0.5, openai_api_key="hello")
    prompt = PromptTemplate.from_template("hello {name}!")
    chain = LLMChain(llm=llm, prompt=prompt)

    import os

    has_env = "OPENAI_API_KEY" in os.environ
    if not has_env:
        os.environ["OPENAI_API_KEY"] = "env_variable"

    llm_2 = OpenAI(model="davinci", temperature=0.5)
    prompt_2 = PromptTemplate.from_template("hello {name}!")
    chain_2 = LLMChain(llm=llm_2, prompt=prompt_2)

    assert dumps(chain_2, pretty=True) == dumps(chain, pretty=True)

    if not has_env:
        del os.environ["OPENAI_API_KEY"]


@pytest.mark.requires("openai")
def test_serialize_llmchain_chat(snapshot: Any) -> None:
    llm = ChatOpenAI(model="davinci", temperature=0.5, openai_api_key="hello")
    prompt = ChatPromptTemplate.from_messages(
        [HumanMessagePromptTemplate.from_template("hello {name}!")]
    )
    chain = LLMChain(llm=llm, prompt=prompt)
    assert dumps(chain, pretty=True) == snapshot

    import os

    has_env = "OPENAI_API_KEY" in os.environ
    if not has_env:
        os.environ["OPENAI_API_KEY"] = "env_variable"

    llm_2 = ChatOpenAI(model="davinci", temperature=0.5)
    prompt_2 = ChatPromptTemplate.from_messages(
        [HumanMessagePromptTemplate.from_template("hello {name}!")]
    )
    chain_2 = LLMChain(llm=llm_2, prompt=prompt_2)

    assert dumps(chain_2, pretty=True) == dumps(chain, pretty=True)

    if not has_env:
        del os.environ["OPENAI_API_KEY"]


@pytest.mark.requires("openai")
def test_serialize_llmchain_with_non_serializable_arg(snapshot: Any) -> None:
    llm = OpenAI(
        model="davinci",
        temperature=0.5,
        openai_api_key="hello",
        client=NotSerializable,
    )
    prompt = PromptTemplate.from_template("hello {name}!")
    chain = LLMChain(llm=llm, prompt=prompt)
    assert dumps(chain, pretty=True) == snapshot
