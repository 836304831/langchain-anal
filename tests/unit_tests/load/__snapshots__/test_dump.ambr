# serializer version: 1
# name: test_person
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "test_dump",
      "Person"
    ],
    "kwargs": {
      "secret": {
        "lc": 1,
        "type": "secret",
        "id": [
          "SECRET"
        ]
      },
      "you_can_see_me": "hello"
    }
  }
  '''
# ---
# name: test_person.1
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "test_dump",
      "SpecialPerson"
    ],
    "kwargs": {
      "another_secret": {
        "lc": 1,
        "type": "secret",
        "id": [
          "ANOTHER_SECRET"
        ]
      },
      "secret": {
        "lc": 1,
        "type": "secret",
        "id": [
          "SECRET"
        ]
      },
      "another_visible": "bye",
      "you_can_see_me": "hello"
    }
  }
  '''
# ---
# name: test_serialize_llmchain
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "langchain",
      "chains",
      "llm",
      "LLMChain"
    ],
    "kwargs": {
      "llm": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "llms",
          "openai",
          "OpenAI"
        ],
        "kwargs": {
          "model": "davinci",
          "temperature": 0.5,
          "openai_api_key": {
            "lc": 1,
            "type": "secret",
            "id": [
              "OPENAI_API_KEY"
            ]
          }
        }
      },
      "prompt": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "prompts",
          "prompt",
          "PromptTemplate"
        ],
        "kwargs": {
          "input_variables": [
            "name"
          ],
          "template": "hello {name}!",
          "template_format": "f-string"
        }
      }
    }
  }
  '''
# ---
# name: test_serialize_llmchain_chat
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "langchain",
      "chains",
      "llm",
      "LLMChain"
    ],
    "kwargs": {
      "llm": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "chat_models",
          "openai",
          "ChatOpenAI"
        ],
        "kwargs": {
          "model": "davinci",
          "temperature": 0.5,
          "openai_api_key": {
            "lc": 1,
            "type": "secret",
            "id": [
              "OPENAI_API_KEY"
            ]
          }
        }
      },
      "prompt": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "prompts",
          "chat",
          "ChatPromptTemplate"
        ],
        "kwargs": {
          "input_variables": [
            "name"
          ],
          "messages": [
            {
              "lc": 1,
              "type": "constructor",
              "id": [
                "langchain",
                "prompts",
                "chat",
                "HumanMessagePromptTemplate"
              ],
              "kwargs": {
                "prompt": {
                  "lc": 1,
                  "type": "constructor",
                  "id": [
                    "langchain",
                    "prompts",
                    "prompt",
                    "PromptTemplate"
                  ],
                  "kwargs": {
                    "input_variables": [
                      "name"
                    ],
                    "template": "hello {name}!",
                    "template_format": "f-string"
                  }
                }
              }
            }
          ]
        }
      }
    }
  }
  '''
# ---
# name: test_serialize_llmchain_with_non_serializable_arg
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "langchain",
      "chains",
      "llm",
      "LLMChain"
    ],
    "kwargs": {
      "llm": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "llms",
          "openai",
          "OpenAI"
        ],
        "kwargs": {
          "model": "davinci",
          "temperature": 0.5,
          "openai_api_key": {
            "lc": 1,
            "type": "secret",
            "id": [
              "OPENAI_API_KEY"
            ]
          },
          "client": {
            "lc": 1,
            "type": "not_implemented",
            "id": [
              "openai",
              "api_resources",
              "completion",
              "Completion"
            ]
          }
        }
      },
      "prompt": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "prompts",
          "prompt",
          "PromptTemplate"
        ],
        "kwargs": {
          "input_variables": [
            "name"
          ],
          "template": "hello {name}!",
          "template_format": "f-string"
        }
      }
    }
  }
  '''
# ---
# name: test_serialize_openai_llm
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "langchain",
      "llms",
      "openai",
      "OpenAI"
    ],
    "kwargs": {
      "model": "davinci",
      "temperature": 0.7,
      "openai_api_key": {
        "lc": 1,
        "type": "secret",
        "id": [
          "OPENAI_API_KEY"
        ]
      }
    }
  }
  '''
# ---
